docs/adapter_impl_notes.md

```diff
@@ -0,0 +1,184 @@
+# Adapter Implementation Notes
+
+This page talks about implementation details of writing adapters. If you want to write your own adapter for your own backend, use this document as a guide.
+
+For this document, we'll use the Gmail Adapter as an example. In this case, the backend consists a set of email messages. Each message contains a timestamp (when the message was received) and fields (email headers such as ``From:``, ``To:``, ``Subject:``, the message body, etc). The Gmail API supports [search expressions](https://support.google.com/mail/answer/7190?hl=en) that select messages based on date, headers, or a free-text search on the message contents.
```

Comment:
Link to Gmail API would be nice. Does the adapter work with a specific version of it? We should call out that an adapter should always document which backend version it is for.